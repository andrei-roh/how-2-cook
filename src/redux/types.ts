export const SET_USER = 'SET_USER';
export const SET_LOADING = 'SET_LOADING';
export const SET_ERROR = 'SET_ERROR';
export const ADD_RECIPES_TO_LIST = 'ADD_RECIPES_TO_LIST';
export const UPDATE_RECIPES_LIST = 'UPDATE_RECIPES_LIST';
export const ADD_IMAGES_TO_LIST = 'ADD_IMAGES_TO_LIST';
export const SET_RECIPE_SEARCH_INPUT = 'SET_RECIPE_SEARCH_INPUT';
export const SET_MANAGE_RECIPES_SCROLL_SIZE = 'SET_MANAGE_RECIPES_SCROLL_SIZE';
export const SET_MANAGE_RECIPES_SCROLL_DIRECTION =
  'SET_MANAGE_RECIPES_SCROLL_DIRECTION';
export const SET_SHOW_MENU = 'SET_SHOW_MENU';
export const ADD_INGREDIENTS_TO_LIST = 'ADD_INGREDIENTS_TO_LIST';
export const SET_INGREDIENT_SEARCH_INPUT = 'SET_INGREDIENT_SEARCH_INPUT';
export const UPDATE_INGREDIENTS_LIST = 'UPDATE_INGREDIENTS_LIST';


