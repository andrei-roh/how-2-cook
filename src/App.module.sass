@use 'src/style/colors'

.app 
  padding-top: 46px
  background-color: colors.$white
  height: 100vh
  box-sizing: border-box
  display: flex
  flex-direction: column
  align-items: center

