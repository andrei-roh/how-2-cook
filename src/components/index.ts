export { Header } from './Header/Header';
export { Input } from './Input/Input';
export { Modal } from './Modal/Modal';
export { Notification } from './Notification/Notification';
export { ImageLoader } from './ImageLoader/ImageLoader';
export { TextArea } from './TextArea/TextArea';
export { Select } from './Select/Select';
export { Link } from './Link/Link';
export { Menu } from './Menu/Menu';
export { RecipeCard } from './RecipeCard/RecipeCard';
