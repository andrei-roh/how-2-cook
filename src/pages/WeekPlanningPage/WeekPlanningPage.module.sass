@use 'src/style/colors'
@use 'src/style/breakpoints'
@import 'src/style/typography'

.weekPlanningPageWrapper 
    width: 100%
    background-color: colors.$white
    box-sizing: border-box
    display: flex
    flex-direction: column
    justify-content: center
    align-items: center
    gap: 8px

    .weekPlanningTitle
        @extend .H3
        padding: 24px
        align-self: flex-start

    .submitButton
        @extend .H6
        width: 147px
        height: 48px
        box-sizing: border-box
        color: colors.$white
        background-color: colors.$blue

        &:hover
            background-color: colors.$sky

        &:disabled,
        &[disabled]
            cursor: auto 
            background-color: colors.$gray

    .recipesBlockWrapper
        width: 100%
        box-sizing: border-box
        padding: 16px
        justify-content: center
        align-items: center