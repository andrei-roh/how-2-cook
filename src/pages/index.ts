export { TitlePage } from './TitlePage/TitlePage';
export { RecipesPage } from './RecipesPage/RecipesPage';
export { CreateRecipePage } from './CreateRecipePage/CreateRecipePage';
export { ShowRecipePage } from './ShowRecipePage/ShowRecipePage';
export { EditRecipePage } from './EditRecipePage/EditRecipePage';
export { IngredientsPage } from './IngredientsPage/IngredientsPage';
export { WeekPlanningPage } from './WeekPlanningPage/WeekPlanningPage';
export { MyPlansPage } from './MyPlansPage/MyPlansPage';