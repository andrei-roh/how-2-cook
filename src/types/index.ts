import { Action } from '@reduxjs/toolkit';

export interface IUser {
  email: string;
}

export enum DishType {
  Main = 'main',
  Soup = 'soup',
  Dessert = 'dessert',
  Salat = 'salat',
  Snack = 'snack',
}

export type RecipeImage = {
  [key: string]: string;
};

export interface IIngredient {
  id: string;
  name: string;
}

export interface IRecipe {
  id: string;
  imageUrl: string;
  name: string;
  type: DishType;
  isVegan: boolean;
  ingredients: string[];
  description: string;
  composition?: string;
  createdAt?: string;
  createdBy?: string;
  updatedAt?: string;
  updatedBy?: string;
}

export interface IPlanningList {
  id: string;
  name: string;
  ingredientsList: string[];
  recipesList: string[];
}

export interface IState {
  user: IUser;
  loading: boolean;
  error: null | string;
  recipesList: IRecipe[];
  imagesList: RecipeImage[];
  recipesPage: {
    recipeSearchInput: string;
    scrollSize: number;
    scrollDirection: ScrollDirection;
  };
  isMenu: boolean;
  ingredientsList: IIngredient[];
  ingredientsPage: {
    ingredientSearchInput: string;
  }
  previousRoute: string;
  planningPage: {
    ingredientsList: string[];
    recipesList: IRecipe[];
  }
}

export interface IAction extends Action {
  type: string;
  payload: never;
}

export interface IValidationError {
  [key: string]: string;
}

export enum FirebaseErrorType {
  InvalidEmail = 'auth/invalid-email',
  InvalidPassword = 'auth/invalid-credential',
  TooManyRequsts = 'auth/too-many-requests',
}

export enum ImageType {
  Webp = 'image/webp',
  Jpeg = 'image/jpeg',
  Jpg = 'image/jpg',
  Png = 'image/png',
}

export interface IOption {
  name: string;
  value: string;
}

export enum ScrollDirection {
  Up = 'up',
  Down = 'down',
}

export enum Severity {
  Success = 'success',
  Info = 'info',
  Warning = 'warning',
  Error = 'error',
}

export interface ICurrentRecipe extends Partial<IRecipe> {}
