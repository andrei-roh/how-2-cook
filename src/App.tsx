import { Outlet } from 'react-router-dom';
import css from './App.module.sass';
import { Header, Notification } from './components';

const App = () => {
  return (
    <>
      <Header />
      <div className={css.app}>
        <Outlet />
      </div>
      <Notification />
    </>
  );
};

export default App;
