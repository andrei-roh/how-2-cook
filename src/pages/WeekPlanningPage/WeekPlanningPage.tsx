import { useCheckAuthentication } from 'src/hooks/useCheckAuthentication';
import css from './WeekPlanningPage.module.sass';
import { useDispatch, useSelector } from 'react-redux';
import { IRecipe, IState } from 'src/types';
import { useCheckAllIngredients } from 'src/hooks/useCheckAllIngredients';
import { useEffect, useState } from 'react';
import { RecipeCard, Select } from 'src/components';
import { getIngredientsToSelect } from 'src/utils/getIngredientsToSelect';
import Button from '@mui/material/Button';
import Stack from '@mui/material/Stack';
import CircularProgress from '@mui/material/CircularProgress';
import { getRecipesByIngredient } from 'src/utils/getRecipesByIngredient';
import { getAllRecipes } from 'src/utils/getAllRecipes';
import {
  addRecipesToList,
  setPreviousRoute,
  updatePlanningIngredientsList,
  updatePlanningRecipesList,
} from 'src/redux/actions';
import { WEEK_PLANNING } from 'src/constants';

export const WeekPlanningPage = () => {
  const dispatch = useDispatch();

  const user = useSelector((state: IState) => state.user);
  const allIngredients = useSelector((state: IState) => state.ingredientsList);
  const planningIngredients = useSelector(
    (state: IState) => state.planningPage.ingredientsList
  );
  const planningRecipes = useSelector(
    (state: IState) => state.planningPage.recipesList
  );
  const currentRecipesList = useSelector((state: IState) => state.recipesList);

  const [recipeIngredients, setRecipeIngredients] = useState<string[] | []>(planningIngredients);
  const [isPlaning, setIsPlaning] = useState(false);
  const [recipes, setRecipes] = useState<IRecipe[] | null>(
    planningRecipes.length > 0 ? planningRecipes : null
  );
  const [isLoading, setIsLoading] = useState(currentRecipesList.length === 0);

  const handleGetRecipesByIngredient = () => {
    setIsPlaning(true);

    (recipeIngredients.length > 0
      ? getRecipesByIngredient(recipeIngredients)
      : getAllRecipes()
    ).then((res) => {
      setRecipes(res);
      setIsPlaning(false);

      dispatch(updatePlanningIngredientsList(recipeIngredients));
      dispatch(updatePlanningRecipesList(res));
      dispatch(setPreviousRoute(WEEK_PLANNING));
    });
  };

  useCheckAuthentication(user);
  useCheckAllIngredients(allIngredients);

  useEffect(() => {
    if (isLoading) {
      getAllRecipes().then((result) => {
        setIsLoading(false);
        dispatch(addRecipesToList(result));
      });
    }
  }, [dispatch, isLoading]);

  return (
    <div className={css.weekPlanningPageWrapper}>
      <div className={css.weekPlanningTitle}>Набор блюд на неделю</div>
      <Select
        value={recipeIngredients}
        setChange={setRecipeIngredients}
        options={getIngredientsToSelect(allIngredients)}
        labelText={'Ингредиенты'}
        multiple
        placeholder={'Выбор'}
        selectClassName={!recipeIngredients ? css.selectPlaceholder : undefined}
      />
      <Button
        className={css.submitButton}
        onClick={handleGetRecipesByIngredient}
      >
        {isPlaning ? (
          <CircularProgress color='inherit' size={32} />
        ) : (
          'Спланировать'
        )}
      </Button>
      {recipes && (
        <Stack
          direction='column'
          useFlexGap
          spacing={2}
          className={css.recipesBlockWrapper}
        >
          {recipes.map((recipe) => (
            <RecipeCard key={recipe.id} {...recipe} />
          ))}
        </Stack>
      )}
    </div>
  );
};
