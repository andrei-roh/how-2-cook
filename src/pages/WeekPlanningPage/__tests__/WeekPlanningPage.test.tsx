import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { DishType, IRecipe, IUser } from 'src/types';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { WeekPlanningPage } from '../WeekPlanningPage';

const navigateMock = vi.fn();
const dispatchMock = vi.fn();
const userMock = { email: 'Test@email.com' };
const imagesListMock = [
  {
    id: 'test_image_src_1',
    value: 'test/image/src/1',
  },
  {
    id: 'test_image_src_2',
    value: 'test/image/src/2',
  },
  {
    id: 'test_image_src_3',
    value: 'test/image/src/3',
  },
];
const recipesListMock: IRecipe[] = [
  {
    id: '00000000-0000-0000-0000-000000000001',
    imageUrl: 'test_image_src_1',
    name: 'Test Recipe 1',
    type: DishType.Main,
    ingredients: [],
    composition: 'Test Ingredients 1',
    description: 'Test Description 1',
    isVegan: false,
  },
  {
    id: '00000000-0000-0000-0000-000000000002',
    imageUrl: 'test_image_src_2',
    name: 'Test Recipe 2',
    type: DishType.Soup,
    ingredients: [],
    composition: 'Test Ingredients 2',
    description: 'Test Description 2',
    isVegan: false,
  },
  {
    id: '00000000-0000-0000-0000-000000000003',
    imageUrl: 'test_image_src_3',
    name: 'Test Recipe 3',
    type: DishType.Salat,
    ingredients: [],
    composition: 'Test Ingredients 3',
    description: 'Test Description 3',
    isVegan: true,
  },
];
const allIngredientsMock = [
  {
    id: '00000000-0000-0000-0000-000000000000',
    name: 'Ingredient 1'
  }
]

const stateMock = {
  user: userMock as IUser,
  loading: false,
  error: null,
  imagesList: imagesListMock,
  recipesList: recipesListMock,
  ingredientsList: allIngredientsMock,
  planningPage: {
    ingredientsList: [],
    recipesList: [],
  }
};

const reducerMock = (state = stateMock) => {
  return state;
};

const storeMock = configureStore({ reducer: reducerMock });

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');

  return {
    ...actual,
    useNavigate: () => ({ navigate: navigateMock }),
  };
});

vi.mock('react-redux', async () => {
  const actual = await vi.importActual('react-redux');

  return {
    ...actual,
    useDispatch: () => ({ dispatch: dispatchMock }),
  };
});

describe('Testing: WeekPlanningPage', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={storeMock}>
        <WeekPlanningPage />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
