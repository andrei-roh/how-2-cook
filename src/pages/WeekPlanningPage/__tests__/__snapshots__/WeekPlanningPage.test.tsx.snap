// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: WeekPlanningPage > should render 1`] = `
<div>
  <div
    class="_weekPlanningPageWrapper_53bfc1"
  >
    <div
      class="_weekPlanningTitle_53bfc1"
    >
      Набор блюд на неделю
    </div>
    <label
      class="_label_9fe140"
    >
      <div>
        Ингредиенты
      </div>
      <div
        class="_select_9fe140"
        tabindex="0"
      >
        Выбор
        <img
          class="_selectArrow_9fe140"
          src="/src/components/Select/assets/arrow.svg"
        />
      </div>
      <div
        class="_selectMenuWrapper_9fe140"
      />
    </label>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary _submitButton_53bfc1 css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Спланировать
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
